File: slt/src/main/java/fr/quentinklein/slt/LocationTracker.java
package fr.quentinklein.slt;

import android.Manifest;
import android.content.Context;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresPermission;
import android.util.Log;

/**
 * @author Quentin Klein <klein.quentin@gmail.com>, Yasir.Ali <ali.yasir0@gmail.com>
 *         <p>
 *         Helper that tracks user location
 *         </p>
 */
public abstract class LocationTracker implements LocationListener {

    /**
     * Tag used for logs
     */
    private static final String TAG = "LocationTracker";
    /**
     * The user location
     * This value is static because, wherever you call a LocationTracker, user location is the same
     */
    private static Location sLocation;

    /**
     * The manager used to track the sLocation
     */
    private LocationManager mLocationManagerService;
    /**
     * Indicates if Tracker is listening for updates or not
     */
    private boolean mIsListening = false;
    /**
     * Indicates if Tracker has found the location or not
     */
    private boolean mIsLocationFound = false;

    /**
     * Any context useful
     */
    private Context mContext;

    /**
     * Settings for the tracker
     */
    private TrackerSettings mTrackerSettings;

    /**
     * Default LocationTracker, uses default values for settings
     *
     * @param context Android context, uiContext is not mandatory.
     */
    @RequiresPermission(anyOf = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION})
    public LocationTracker(@NonNull Context context) {
        this(context, TrackerSettings.DEFAULT);
    }

    /**
     * Customized LocationTracker, uses the specified services and starts listening for a location.
     *
     * @param context         Android context, uiContext is not mandatory.
     * @param trackerSettings {@link TrackerSettings}, the tracker settings
     */
    @RequiresPermission(anyOf = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION})
    public LocationTracker(@NonNull Context context, @NonNull TrackerSettings trackerSettings) {
        this.mContext = context;
        this.mTrackerSettings = trackerSettings;

        // Get the location manager
        this.mLocationManagerService = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        // default
        if (sLocation == null && trackerSettings.shouldUseGPS()) {
            LocationTracker.sLocation = mLocationManagerService.getLastKnownLocation(LocationManager.GPS_PROVIDER);
        }
        if (sLocation == null && trackerSettings.shouldUseNetwork()) {
            LocationTracker.sLocation = mLocationManagerService.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
        }
        if (sLocation == null && trackerSettings.shouldUsePassive()) {
            LocationTracker.sLocation = mLocationManagerService.getLastKnownLocation(LocationManager.PASSIVE_PROVIDER);
        }
    }

    /**
     * Deprecated since 3.1
     * Used to make the tracker listening for location updates
     *
     * @see #startListening()
     */
    @Deprecated
    public final void startListen() {
        startListening();
    }

    /**
     * Make the tracker listening for location updates
     */
    @RequiresPermission(anyOf = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION})
    public final void startListening() {
        if (!mIsListening) {
            Log.i(TAG, "LocationTracked is now listening for location updates");
            // Listen for GPS Updates
            if (mTrackerSettings.shouldUseGPS()) {
                if (LocationUtils.isGpsProviderEnabled(mContext)) {
                    mLocationManagerService.requestLocationUpdates(LocationManager.GPS_PROVIDER, mTrackerSettings.getTimeBetweenUpdates(), mTrackerSettings.getMetersBetweenUpdates(), this);
                } else {
                    onProviderError(new ProviderError(LocationManager.GPS_PROVIDER, "Provider is not enabled"));
                }
            }
            // Listen for Network Updates
            if (mTrackerSettings.shouldUseNetwork()) {
                if (LocationUtils.isNetworkProviderEnabled(mContext)) {
                    mLocationManagerService.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, mTrackerSettings.getTimeBetweenUpdates(), mTrackerSettings.getMetersBetweenUpdates(), this);
                } else {
                    onProviderError(new ProviderError(LocationManager.NETWORK_PROVIDER, "Provider is not enabled"));
                }
            }
            // Listen for Passive Updates
            if (mTrackerSettings.shouldUsePassive()) {
                if (LocationUtils.isPassiveProviderEnabled(mContext)) {
                    mLocationManagerService.requestLocationUpdates(LocationManager.PASSIVE_PROVIDER, mTrackerSettings.getTimeBetweenUpdates(), this.mTrackerSettings.getMetersBetweenUpdates(), this);
                } else {
                    onProviderError(new ProviderError(LocationManager.PASSIVE_PROVIDER, "Provider is not enabled"));
                }
            }
            mIsListening = true;

            // If user has set a timeout
            if (mTrackerSettings.getTimeout() != -1) {
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        if (!mIsLocationFound && mIsListening) {
                            Log.i(TAG, "No location found in the meantime");
                            stopListening();
                            onTimeout();
                        }
                    }
                }, mTrackerSettings.getTimeout());
            }
        } else {
            Log.i(TAG, "Relax, LocationTracked is already listening for location updates");
        }
    }

    /**
     * Deprecated since 3.1
     * Used to make the tracker stops listening for location updates
     *
     * @see #stopListening()
     */
    @Deprecated
    public final void stopListen() {
        stopListening();
    }

    /**
     * Make the tracker stops listening for location updates
     */
    public final void stopListening() {
        if (mIsListening) {
            Log.i(TAG, "LocationTracked has stopped listening for location updates");
            mLocationManagerService.removeUpdates(this);
            mIsListening = false;
        } else {
            Log.i(TAG, "LocationTracked wasn't listening for location updates anyway");
        }
    }

    /**
     * Best effort, it calls {@link #onLocationChanged(Location)} with static field named {@link #sLocation} if it is not null
     */
    public final void quickFix() {
        if (LocationTracker.sLocation != null) {
            onLocationChanged(LocationTracker.sLocation);
        }
    }

    /**
     * Getter used to know if the Tracker is listening at this time.
     *
     * @return true if the tracker is listening, false otherwise
     */
    public final boolean isListening() {
        return mIsListening;
    }

    /**
     * Called when the tracker had found a location
     *
     * @see android.location.LocationListener#onLocationChanged(android.location.Location)
     */
    @Override
    public final void onLocationChanged(@NonNull Location location) {
        Log.i(TAG, "Location has changed, new location is " + location);
        LocationTracker.sLocation = new Location(location);
        mIsLocationFound = true;
        onLocationFound(location);
    }

    /**
     * Called when the tracker had found a location
     *
     * @param location the found location
     */
    public abstract void onLocationFound(@NonNull Location location);

    /**
     * Called when the tracker had not found any location and the timeout just happened
     */
    public abstract void onTimeout();

    /**
     * Called when a provider has been disabled.
     * By default, this method do nothing but a Log on i
     *
     * @see android.location.LocationListener#onProviderDisabled(java.lang.String)
     */
    @Override
    public void onProviderDisabled(@NonNull String provider) {
        // By default do nothing but log
        Log.i(TAG, "Provider (" + provider + ") has been disabled");
    }

    /**
     * Called when a provider has been enabled.
     * By default, this method do nothing but a Log on i
     *
     * @see android.location.LocationListener#onProviderEnabled(java.lang.String)
     */
    @Override
    public void onProviderEnabled(@NonNull String provider) {
        // By default do nothing but log
        Log.i(TAG, "Provider (" + provider + ") has been enabled");
    }

    /**
     * Called when status has changed.
     * By default, this method do nothing but a Log on i
     *
     * @see android.location.LocationListener#onStatusChanged(java.lang.String, int, android.os.Bundle)
     */
    @Override
    public void onStatusChanged(@NonNull String provider, int status, Bundle extras) {
        // By default do nothing but log
        Log.i(TAG, "Provider (" + provider + ") status has changed, new status is " + status);
    }

    /**
     * Triggered when there was an error with a provider, most of the time, when this one is not enabled
     *
     * @param providerError the provider error
     */
    public void onProviderError(@NonNull ProviderError providerError) {
        // By default do nothing but log
        Log.w(TAG, "Provider (" + providerError.getProvider() + ")", providerError);
    }

}

File: slt/src/main/java/fr/quentinklein/slt/LocationTracker.kt
package fr.quentinklein.slt

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Build
import android.os.Bundle
import androidx.annotation.NonNull
import androidx.annotation.RequiresPermission
import android.util.Log

/**
 * Build a LocationTracker
 * @param minTimeBetweenUpdates The minimum time interval between location updates, in milliseconds by default its value is 5 minutes
 * @param minDistanceBetweenUpdates The minimum distance between location updates in meters, by default its value is 100m
 * @param shouldUseGPS Specifies if tracker should use the GPS (default is true)
 * @param shouldUseNetwork Specifies if tracker should use the Network (default is true)
 * @param shouldUsePassive Specifies if tracker should use the Passive provider (default is true)
 */
class LocationTracker constructor(
        val minTimeBetweenUpdates: Long = 5 * 60 * 1000.toLong(),
        val minDistanceBetweenUpdates: Float = 100f,
        val shouldUseGPS: Boolean = true,
        val shouldUseNetwork: Boolean = true,
        val shouldUsePassive: Boolean = true
) {
    // Android LocationManager
    private lateinit var locationManager: LocationManager

    // Last known location
    private var lastKnownLocation: Location? = null

    // Custom Listener for the LocationManager
    private val listener = object : LocationListener {
        override fun onLocationChanged(p0: Location) {
            Location(p0).let { currentLocation ->
                lastKnownLocation = currentLocation
                hasLocationFound = true
                listeners.forEach { l -> l.onLocationFound(currentLocation) }
            }
        }

        override fun onProviderDisabled(provider: String) {
            behaviorListener.forEach { l -> l.onProviderDisabled(provider) }
        }

        override fun onProviderEnabled(provider: String) {
            behaviorListener.forEach { l -> l.onProviderEnabled(provider) }
        }

        override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {
            behaviorListener.forEach { l -> l.onStatusChanged(provider, status, extras) }
        }

    }

    // List used to register the listeners to notify
    private val listeners: MutableSet<Listener> = mutableSetOf()

    // List used to register the behavior listeners to notify
    private val behaviorListener: MutableSet<BehaviorListener> = mutableSetOf()

    /**
     * Indicates if Tracker is listening for updates or not
     */
    var isListening = false
        private set

    /**
     * Indicates if Tracker has found the location or not
     */
    var hasLocationFound = false
        private set

    /**
     * Add a listener to the stack so it will be notified once a new location is found
     * @param listener the listener to add to the list.
     * @return true if the listener has been added, false otherwise
     */
    fun addListener(listener: Listener): Boolean = listeners.add(listener)

    /**
     * Remove a listener from the stack
     * @param listener the listener to remove from the list.
     * @return true if the listener has been removed, false otherwise
     */
    fun removeListener(listener: Listener): Boolean = listeners.remove(listener)

    /**
     * Add a behavior listener to the stack so it will be notified when a provider is updated
     * @param listener the listener to add to the list.
     * @return true if the listener has been added, false otherwise
     */
    fun addBehaviorListener(listener: BehaviorListener): Boolean = behaviorListener.add(listener)

    /**
     * Remove a behavior listener from the stack
     * @param listener the listener to remove from the list.
     * @return true if the listener has been removed, false otherwise
     */
    fun removeBehaviorListener(listener: BehaviorListener): Boolean = behaviorListener.remove(listener)


    /**
     * Make the tracker listening for location updates
     * @param context a valid android.content.Context
     */
    @RequiresPermission(anyOf = [Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION])
    fun startListening(context: Context) {
        initManagerAndUpdateLastKnownLocation(context)
        if (!isListening) {
            // Listen for GPS Updates
            if (shouldUseGPS) {
                registerForLocationUpdates(LocationManager.GPS_PROVIDER)
            }
            // Listen for Network Updates
            if (shouldUseNetwork) {
                registerForLocationUpdates(LocationManager.NETWORK_PROVIDER)
            }
            // Listen for Passive Updates
            if (shouldUseNetwork) {
                registerForLocationUpdates(LocationManager.PASSIVE_PROVIDER)
            }
            isListening = true
        }
    }

    /**
     * Make the tracker stops listening for location updates
     * @param clearListeners optional (default false) drop all the listeners if set to true
     */
    fun stopListening(clearListeners: Boolean = false) {
        if (isListening) {
            locationManager.removeUpdates(listener)
            isListening = false
            if (clearListeners) {
                listeners.clear()
            }
        }
    }

    /**
     * Best effort, it calls [.onLocationChanged] with static field named [.lastKnownLocation] if it is not null
     * @param context Context
     */
    @RequiresPermission(anyOf = [Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION])
    fun quickFix(@NonNull context: Context) {
        initManagerAndUpdateLastKnownLocation(context)
        lastKnownLocation?.let { lastLocation ->
            listeners.forEach { l -> l.onLocationFound(lastLocation) }
        }
    }

    @SuppressLint("MissingPermission")
    private fun initManagerAndUpdateLastKnownLocation(context: Context) {
        // Init the manager
        locationManager = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            context.getSystemService(LocationManager::class.java)
        } else {
            context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
        }
        // Update the lastKnownLocation
        if (lastKnownLocation == null && shouldUseGPS) {
            lastKnownLocation = locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER)
        }
        if (lastKnownLocation == null && shouldUseNetwork) {
            lastKnownLocation = locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)
        }
        if (lastKnownLocation == null && shouldUsePassive) {
            lastKnownLocation = locationManager.getLastKnownLocation(LocationManager.PASSIVE_PROVIDER)
        }
    }

    @SuppressLint("MissingPermission")
    private fun registerForLocationUpdates(provider: String) {
        if (locationManager.isProviderEnabled(provider)) {
            locationManager.requestLocationUpdates(provider, minTimeBetweenUpdates, minDistanceBetweenUpdates, listener)
        } else {
            listeners.forEach { l -> l.onProviderError(ProviderError("Provider `$provider` is not enabled")) }
        }
    }

    interface Listener {
        /**
         * Called when the tracker had found a location
         *
         * @param location the found location
         */
        fun onLocationFound(location: Location)

        /**
         * Called when there is an error on a specific provider
         * @param providerError the error sent
         */
        fun onProviderError(providerError: ProviderError)
    }

    interface BehaviorListener {
        /**
         * See android.location.LocationListener#onProviderDisabled
         */
        fun onProviderDisabled(provider: String)

        /**
         * See android.location.LocationListener#onProviderEnabled
         */
        fun onProviderEnabled(provider: String)

        /**
         * See android.location.LocationListener#onStatusChanged
         */
        fun onStatusChanged(provider: String, status: Int, extras: Bundle)
    }

}
